File: two_pointers/problems/js/rotateList.js
class ListNode {
  constructor(val = 0, next = null) {
    this.val = val; // Number
    this.next = next; // ListNode | null
  }
}


function rotateList(head,k){
  if (!head ||!head.next || k === 0) return head;

  // count the number of nodes in the head and find the tail node
  let count = 1;
  let tail = head;
  while (tail.next) {
    tail = tail.next;
    count++;
  }

  // adjust the value of k to be within the range of 0 to count - 1
  k = k % count;
  if (k === 0) return head;

  // find the new tail node
  let newTailPosition = n - k - 1
  let newTail = head;
  for (let i = 0; i < newTailPosition; i++) {
    newTail = newTail.next;
  }

  // perform the rotate operation

  const newHead = newTail.next
  newTail.next = null
  tail.next = head

  return newHead
}

File: two_pointers/problems/python/rotateList.py
def rotateRight(head: Optional[ListNode], k: int) -> Optional[ListNode]:
    if not head or not head.next or k == 0:
        return head

    # Count the number of nodes and find the tail
    n = 1
    tail = head
    while tail.next:
        tail = tail.next
        n += 1

    # Adjust k to avoid unnecessary rotations
    k %= n
    if k == 0:
        return head

    # Find the new tail: (n - k - 1)th node
    new_tail_position = n - k - 1
    new_tail = head
    for _ in range(new_tail_position):
        new_tail = new_tail.next

    # Perform the rotation
    new_head = new_tail.next
    new_tail.next = None
    tail.next = head

    return new_head
